### PR TITLE
Full JetEngine Support

### PR DESCRIPTION
**Pull request by rwitten**

### CODE REVIEW

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -500,7 +505,7 @@ def kv_cache_prefill(self,
       """
       batch, sequence, heads, kv_head_size = key.shape
       assert key.dtype == value.dtype, "Key and Value Dtypes should match."
-      assert self.max_prefill_predict_length == sequence, "Set prefill length must match prefill sequence"
+      #assert self.max_prefill_predict_length == sequence, "Set prefill length must match prefill sequence {self.max_prefill_predict_length=}, {sequence=}"
```
Comment (patemotter) [inline MaxText/layers/attentions.py:508]: Delete if no longer needed as an assert.

File: MaxText/maxengine_server.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+"""Runs a server with maxtext."""
+
+from typing import Sequence
+
+from absl import app
+from absl import flags
+import jax
+import os
+import sys
+import pyconfig # change the path to maxtext pyconfig
```
Comment (patemotter) [inline MaxText/maxengine_server.py:10]: Does this need to be done before merge?

File: MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -88,3 +88,33 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
   in_tree, out_tree = get_train_input_output_trees(partial_train, shaped_input_args, shaped_input_kwargs)
   p_train_step = deserialize_and_load(serialized_compiled, in_tree, out_tree)
   return p_train_step
+
+# https://arxiv.org/pdf/2204.02311.pdf Appendix B
+def calculate_tflops_training_per_device(num_model_parameters, config):
+  """ Calculate training TFLOP"""
+  learnable_weight_tflops = 6 * num_model_parameters * config.max_target_length * config.per_device_batch_size \
+                                   / 10**12
+  noncasual_attention_flops = 12 * config.num_query_heads * config.num_decoder_layers * config.head_dim \
```
Comment (patemotter) [inline MaxText/maxtext_utils.py:97]: Nit: `casual` -> `causal`

**Review by patemotter — APPROVED** [review]
LGTM, very minor comments.

File: MaxText/maxengine_config.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+from typing import Type
+import functools
+
+import dataclasses
+import jax
+from typing import Any, Callable, Tuple, Type, Union, cast
+
+from jetstream.core import config_lib
+from jetstream.engine import engine_api
+import maxengine
```
Comment (qihqi) [inline MaxText/maxengine_config.py:11]: Q. this import is suppose to load a `maxengine.py` file in the same directory? It seems to be missing

**Review by patemotter — APPROVED** [review]
LGTM except for the pylint check failing.

File: MaxText/maxengine.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
+
```
Comment (vipannalla) [inline MaxText/maxengine.py:1]: Can we move this file, "maxengine_server.py", and "maxengine_server.py" to its own sub-directory ( "jetstream server" or some better name) to have a clean separation of jetstream integration code?

File: MaxText/maxengine_server.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+import  maxengine_config
+from jetstream.core import server_lib
```
Comment (vipannalla) [inline MaxText/maxengine_server.py:13]: JetStream need to be added in requirements.txt, I don't see it in there rn. Is that auto added through a different mechanism?

Comment (rwitten) [reply MaxText/maxengine_server.py:13]: needs to be added

File: MaxText/configs/base.yml [hunk 7]
```diff
@@ -86,7 +86,8 @@ logits_dot_in_fp32: True  # whether to use fp32 in logits_dense or shared_embedd
 remat_policy: 'full'
 scan_layers: True
 param_scan_axis: 1
-attention: 'flash' # Supported attention: dot_product, flash, cudnn_flash_te
+attention: 'recommended' # Supported attention: recommended, dot_product, flash, cudnn_flash_te
```
[the remark this answers is not in the document]
Comment (rwitten) [reply MaxText/configs/base.yml:89]: Can you suggest a better self documenting name? I don't think a comment serves our interest.

Context: "recommended" chooses the kernel that best fits your need based on MaxText's guess.

[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/configs/base.yml:89]: How about "auto_select" or "choose_best_fit"?

[the remark this answers is not in the document]
Comment (rwitten) [reply MaxText/configs/base.yml:89]: I decided on `autoselected` thanks

File: MaxText/maxengine.py (hunk 5 above)
Comment (rwitten) [reply MaxText/maxengine.py:1]: I'm not going to worry about this right now (I'm already the bottleneck for too much, things are simple enough and it isn't a one way door) but I'm happy for someone to take this on in the future.

File: MaxText/maxengine_server.py (hunk 2 above)
Comment (rwitten) [reply MaxText/maxengine_server.py:10]: updated comment.

File: MaxText/layers/attentions.py (hunk 1 above)
Comment (rwitten) [reply MaxText/layers/attentions.py:508]: thanks deleted

File: MaxText/inference_analysis/analyze_sharegpt.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import json
+
+CUTOFF_INPUT = 1024
+CUTOFF_OUTPUT = 1024
+
+prefill_bucket_size_to_ms = {64 : 14.02, 128:18.29, 256:23.59, 512:35.28, 1024: 60.28}
```
[the remark this answers is not in the document]
Comment (rwitten) [reply MaxText/inference_analysis/analyze_sharegpt.py:6]: Made a better directory name

**Pull request merged**